### PR TITLE
removenode: allow specifying nodes to ignore using host_id

### PR DESCRIPTION
**Pull request by bhalevy**
Currently, when specifying nodes to ignore for replace or removenode,
we support specifying them only using their ip address.

As discussed in https://github.com/scylladb/scylladb/issues/11839 for removenode,
we intentionally require the host uuid for specifying the node to remove,
so the nodes to ignore (that are also done, otherwise we need not ignore them),
should be consistent with that and be specified using their host_id.

The series extends the apis and allows either the nodes ip address or their host_id
to be specified, for backward compatibility.

We should deprecate the ip address method over time and convert the tests and management
software to use the ignored nodes' host_id:s instead.

### CODE REVIEW

**Comment by bhalevy** [comment]
cc @annastuchlik for the docs changes in the series.

**Comment by bhalevy** [comment]
Note, this PR is ready for review, but I converted it to a draft since
it depends on https://github.com/scylladb/scylladb/pull/11840 which I separated out to expedite it.

File: api/storage_service.cc
Comment (asias) [inline api/storage_service.cc:720]: I understand why you convert to string here. But it is sad. Can we do the full parse of  the ignore_nodes here?

File: locator/token_metadata.hh
Comment (asias) [inline locator/token_metadata.hh:225]: It is better to have a 
struct inet_adddres_or_host_id {
}

File: service/storage_service.cc
Comment (asias) [inline service/storage_service.cc:2345]: If we parse the host_id in api/, we can keep the strong type for ignore_nodes parameter.

File: api/api-doc/storage_service.json
Comment (asias) [inline api/api-doc/storage_service.json:1248]: What if user passes a mix of host id and ip address, can we detect and reject this?

**Comment by asias** [comment]
Benny, I like what is achieved in this PR. Let's figure out some thing better than mixing host id an ip address in string though.

File: api/api-doc/storage_service.json
Comment (bhalevy) [reply api/api-doc/storage_service.json:1248]: Why do we care?
I think that as a follow-up we should deprecate the use of ip-addresses for ignore nodes, and eventually we'll support only host IDs.

File: service/storage_service.cc
Comment (bhalevy) [reply service/storage_service.cc:2345]: what about the host_id? it's also passed as a string so we should do all the parsing in the api layer or here.
I prefer to do it here since we have the token_metadata at hand, while the api layer doesn't know about it and it's current version.

File: api/storage_service.cc
Comment (asias) [reply api/storage_service.cc:720]: Because we know the host_id in the string is for sure host_id. But before this patch, we pass std::list<gms::inet_address> to ss:removenode(). After, we pass a std::list<sstring>. This is a step back.  We lose the type info. If after this patch, we must to do the parse in removenode, let's do all the parse there, i.e., moving the parse from ss::remove_node.set to ss:removenode. Not part of the parse in api and part in stroage_service.

File: api/api-doc/storage_service.json
Comment (asias) [reply api/api-doc/storage_service.json:1248]: This is a user api change. We care because we need to make sure wrong user inputs are handled. Do we handle it correct?

Comment (asias) [reply api/api-doc/storage_service.json:1248]: Mixing the host id and ip are most probably an operation mistake. I think it is better just to reject such user input. The list has to be either all host id or all ips.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:2345]: We can convert the host_id to uuid and pass too. But  see my comments above, we can do the parse in token_metada.

File: api/storage_service.cc
Comment (asias) [reply api/storage_service.cc:720]: We could verify the host id is at least a uuid in the api layer. 
But see my comments above, move all the parse stroage_sercie:: removenode().  

Do you agree it is better to do parse in one place?

Comment (bhalevy) [reply api/storage_service.cc:720]: In general, I think the api layer should be responsible for the superficial parsing and validating the syntax, rejecting invalid arguments, while the semantic validation should be done at the service layer.
I.e. it's out of scope of the service::storage_service layer to make sure the list of nodes to ignore is all UUIDs or all ip addresses - that's the api's business.

We can change the storage_service api to get strongly typed parameters and do further validation based on the token_metadata, but I think it's a hassle, and in this case it's fine to pass strings down the stack.

In the next version, I'll some basic validation at the api layer so the strings are at least validated from the syntax perspective.

**Comment by bhalevy** [comment]
In v2 ([c8886e1](https://github.com/scylladb/scylladb/commit/c8886e1fce4008f0b70b20ed71a25f033d14580a)):
- define and use locator::host_id_or_endpoint
- validate remove_node parameters syntax in api layer
  - pass strongly typed params to storage_service::removenode
  - do the semantic validation at the service layer by resolving hsot_id to endpoint or vice versa

**Comment by asias** [comment]
v2 looks good. Thanks.

**Comment by denesb** [comment]
@bhalevy it looks like the dependency has been merged. Please rebase and take it out of Draft if it is ready for merging.

File: locator/token_metadata.hh
Comment (denesb) [inline locator/token_metadata.hh:138]: This should maybe be in api/, if the only user of it is there.

Comment (bhalevy) [reply locator/token_metadata.hh:138]: storage_service needs it to.
I think that since the host id and endpoint are location information this structure should be in the locator namespace. maybe in a separate header file, but we can extract it later to.

In the future, when we use either the endpoint or the host_id to identify nodes, we'll need a registry and a `class node` abstraction.  The mapping registry today is effectively held in the token metadata.

File: api/storage_service.cc
Comment (denesb) [inline api/storage_service.cc:737]: This can be condensed into a single if:
```
if (hoep.has_host_id() != ignore.nodes.front().has_host_id()) {
```

File: locator/token_metadata.hh
Comment (bhalevy) [reply locator/token_metadata.hh:138]: Maybe we should extract stuff to locator/topology.hh as a follow up.

File: api/api-doc/storage_service.json
Comment (denesb) [inline api/api-doc/storage_service.json:1231]: s/remapped and restored onto/redistributed among/

Why the `if possible`? Is there a situation under this is dong possible? A failure is always implicit I think, no need to include it in the message.

File: locator/token_metadata.hh
Comment (denesb) [reply locator/token_metadata.hh:138]: If it has more users, I'm fine with it staying here.

File: api/storage_service.cc
Comment (bhalevy) [reply api/storage_service.cc:737]: I wanted the error to be more user friendly about what is expected, but maybe it's not that important, after all the list should be symmetrical.

File: api/api-doc/storage_service.json
Comment (bhalevy) [reply api/api-doc/storage_service.json:1231]: I think we can be in a degraded situation where we cannot satisfy the replication factor if removing a node when RF == number of nodes.

**Comment by bhalevy** [comment]
In v3 [335a8cc](https://github.com/scylladb/scylladb/commit/335a8cc36286411f224228cd441b865b22ea9f01)):
- rebased + retested
- simplified `hoep.has_host_id() != ignore.nodes.front().has_host_id()` condition and error message
- `s/remapped and restored onto/redistributed among/` in remove_node json description

File: api/storage_service.cc
Comment (bhalevy) [reply api/storage_service.cc]: done in v3

File: api/api-doc/storage_service.json
Comment (bhalevy) [reply api/api-doc/storage_service.json]: done in v3

**Review by denesb — APPROVED** [review]
LGTM, will queue once CI approves.

**Pull request closed without merging**
